File: eBay/165.compare-version-numbers.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    int compareVersion(string version1, string version2) {
        /*
        Approach: parse integer revisions separated by dots from both strings.
        Missing trailing revisions are treated as zero. Return at the first
        unequal revision.

        Complexity: O(m+n) time, O(1) extra space.
        */
        int i = 0, j = 0;
        while (i < (int)version1.size() || j < (int)version2.size()) {
            long long a = 0, b = 0;
            while (i < (int)version1.size() && version1[i] != '.') a = a * 10 + version1[i++] - '0';
            while (j < (int)version2.size() && version2[j] != '.') b = b * 10 + version2[j++] - '0';
            if (a < b) return -1;
            if (a > b) return 1;
            if (i < (int)version1.size()) ++i;
            if (j < (int)version2.size()) ++j;
        }
        return 0;
    }
};

File: pt-4/165.compare-version-numbers.go
package leetcode

import (
	"strconv"
	"strings"
)

// CompareVersion165 compares dot-separated integer revisions, treating missing
// trailing revisions as zero. strconv.Atoi naturally ignores leading zeros.
//
// Time: O(m+n). Space: O(m+n) for split parts.
func CompareVersion165(version1 string, version2 string) int {
	a, b := strings.Split(version1, "."), strings.Split(version2, ".")
	max := len(a)
	if len(b) > max {
		max = len(b)
	}
	for i := 0; i < max; i++ {
		x, y := 0, 0
		if i < len(a) {
			x, _ = strconv.Atoi(a[i])
		}
		if i < len(b) {
			y, _ = strconv.Atoi(b[i])
		}
		if x < y {
			return -1
		}
		if x > y {
			return 1
		}
	}
	return 0
}
